### PR TITLE
refactor!: Use enum class for NavigationDirection

### PR DESCRIPTION
**Pull request by paulgessinger**
This change is motivated by arithmetic between non-class enums and other types that are not the same enum is deprecated in C++20. I overload the operators we use these enums in to support the same use cases.

I think this is breaking since it's part of the public propagator interface.

BREAKING CHANGE: The `NavigationDirection` enum changes to `enum class`. The values are renamed `NavigationDiretcion::forward` -> `NavigationDirection::Forward` and `NavigationDirection::backward` -> `NavigationDirection::Backward`.

### CODE REVIEW

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ struct KalmanFitterOptions {
   /// @param mScattering Whether to include multiple scattering
   /// @param eLoss Whether to include energy loss
   /// @param rFiltering Whether to run filtering in reversed direction as smoothing
-  /// @param rfScaling Scale factor for the covariance matrix before the backward filtering
+  /// @param rfScaling Scale factor for the covariance matrix before the NavigationDirection::Backward filtering
```
Comment (stephenswat) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:108]: This is probably an unintended find-replace artifact.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 2]
```diff
@@ -206,7 +206,8 @@ struct KalmanFitterResult {
   // Measurement surfaces without hits
   std::vector<const Surface*> missedActiveSurfaces;
 
-  // Measurement surfaces handled in both forward and backward filtering
+  // Measurement surfaces handled in both NavigationDirection::Forward and
+  // NavigationDirection::Backward filtering
```
Comment (stephenswat) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:210]: Same here, bad replace.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 3]
```diff
@@ -319,7 +320,9 @@ class KalmanFitter {
       // - Waiting for a current surface
       auto surface = state.navigation.currentSurface;
       std::string direction =
-          (state.stepping.navDir == forward) ? "forward" : "backward";
+          (state.stepping.navDir == NavigationDirection::Forward)
+              ? "NavigationDirection::Forward"
+              : "NavigationDirection::Backward";
```
Comment (stephenswat) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:325]: Same here.

File: Core/src/Definitions/Common.cpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  switch (navDir) {
+    case NavigationDirection::Forward:
+      os << "forward";
+      break;
+    case NavigationDirection::Backward:
+      os << "backward";
+      break;
+    default:
+      assert(false && "Invalid direction (shouldn't happen)");
+      std::abort();
```
Comment (stephenswat) [inline Core/src/Definitions/Common.cpp:25]: Maybe add a `__builtin_unreachable` here?

File: Core/src/Geometry/Layer.cpp [hunk 5]
```diff
@@ -105,7 +105,7 @@ Acts::Layer::compatibleSurfaces(
   boost::container::small_vector<SurfaceIntersection, 10> sIntersections;
 
   // fast exit - there is nothing to
-  if (!m_surfaceArray || !m_approachDescriptor || !options.navDir) {
+  if (!m_surfaceArray || !m_approachDescriptor) {
```
Comment (stephenswat) [inline Core/src/Geometry/Layer.cpp:108]: Shouldn't this be replaced by `|| options.navDir == Forward`?

File: Core/src/Geometry/Layer.cpp [hunk 6]
```diff
@@ -180,7 +180,7 @@ Acts::Layer::compatibleSurfaces(
     if (sfi && detail::checkIntersection(sfi.intersection, pathLimit,
                                          overstepLimit, s_onSurfaceTolerance)) {
       // Now put the right sign on it
-      sfi.intersection.pathLength *= std::copysign(1., options.navDir);
+      sfi.intersection.pathLength *= options.navDir;
```
Comment (stephenswat) [inline Core/src/Geometry/Layer.cpp:183]: Doesn't this need to be explicitly converted to an integer (or underlying type) in this case?

File: Core/src/Geometry/TrackingVolume.cpp [hunk 7]
```diff
@@ -480,8 +482,7 @@ Acts::TrackingVolume::compatibleBoundaries(
                  << bSurface->surfaceRepresentation().geometryId());
     if (detail::checkIntersection(sIntersection.intersection, pLimit, oLimit,
                                   s_onSurfaceTolerance, logger)) {
-      sIntersection.intersection.pathLength *=
-          std::copysign(1., options.navDir);
+      sIntersection.intersection.pathLength *= options.navDir;
```
Comment (stephenswat) [inline Core/src/Geometry/TrackingVolume.cpp:485]: Why can we skip the conversion here?

File: Core/src/Geometry/TrackingVolume.cpp [hunk 8]
```diff
@@ -490,8 +491,7 @@ Acts::TrackingVolume::compatibleBoundaries(
       ACTS_VERBOSE("Consider alternative");
       if (detail::checkIntersection(sIntersection.alternative, pLimit, oLimit,
                                     s_onSurfaceTolerance, logger)) {
-        sIntersection.alternative.pathLength *=
-            std::copysign(1., options.navDir);
+        sIntersection.alternative.pathLength *= options.navDir;
```
Comment (stephenswat) [inline Core/src/Geometry/TrackingVolume.cpp:494]: And here?

File: Core/src/Geometry/TrackingVolume.cpp [hunk 9]
```diff
@@ -656,7 +656,7 @@ Acts::TrackingVolume::compatibleSurfacesFromHierarchy(
   double pLimit = options.pathLimit;
   double oLimit = options.overstepLimit;
 
-  if (m_bvhTop == nullptr || !options.navDir) {
+  if (m_bvhTop == nullptr) {
```
Comment (stephenswat) [inline Core/src/Geometry/TrackingVolume.cpp:659]: Same comment as before, is this valid?

**Review by stephenswat — CHANGES_REQUESTED** [review]
Massive support for going to type-safe enums here, just a few comments where I think find-replace might have gone wrong.

File: Core/src/Geometry/TrackingVolume.cpp (hunk 9 above)
Comment (stephenswat) [reply Core/src/Geometry/TrackingVolume.cpp:659]: Never mind, I see how this works now.

File: Core/src/Geometry/Layer.cpp (hunk 5 above)
Comment (stephenswat) [reply Core/src/Geometry/Layer.cpp:108]: No, it shouldn't -- never mind. Why can't I resolve my own comments?

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp:108]: oh, yes, you're right Stephen!

**Review by asalzburger — DISMISSED** [review]
There have been some replace errors that can/should be fixed.

File: Core/src/Geometry/Layer.cpp (hunk 6 above)
Comment (paulgessinger) [reply Core/src/Geometry/Layer.cpp:183]: I defined the operators for these types.

File: Core/src/Definitions/Common.cpp (hunk 4 above)
Comment (paulgessinger) [reply Core/src/Definitions/Common.cpp:25]: For string output, so you think it's necessary? Would also have to be behind a compiler aware macro, right?

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 3 above)
Comment (paulgessinger) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp]: Should be fixed now.

**Comment by paulgessinger** [comment]
I updated the branch and fixed the bad replaces.

**Pull request merged**